R/foundry_sql_client.R

```diff
@@ -0,0 +1,222 @@
+
+      resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, url_path),
+                                      method = "GET",
+                                      query_params = query_params,
+                                      header_params = header_params,
+                                      body = body)
+
+      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
+        deserialized_resp_obj <- tryCatch(
+          jsonlite::fromJSON(httr::content(resp, "text", encoding = "UTF-8")),
+          error = function(e) {
+            stop("Failed to deserialize response")
+          }
+        )
+        ApiResponse$new(deserialized_resp_obj, resp)
+      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 399) {
+        ApiResponse$new(paste("Server returned ", httr::status_code(resp), " response status code."), resp)
+      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
+        ApiResponse$new("API client error", resp)
+      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
+        ApiResponse$new("API server error", resp)
+      }
+    },
+
+    get_results = function(query_id) {
+      api_response <- self$get_results_with_http_info(query_id)
+      resp <- api_response$response
+      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
+        api_response$content
+      } else if (httr::status_code(resp) >= 300 && httr::status_code(resp) <= 599) {
+        httr::stop_for_status(resp)
+      }
+    },
+    get_results_with_http_info = function(query_id) {
+      query_params <- list()
+      header_params <- c()
+
+      if (missing(`query_id`)) {
+        stop("Missing required parameter `query_id`.")
+      }
+
+      body <- NULL
+      url_path <- "/{queryId}/results"
+      url_path <- gsub(paste0("\\{", "queryId", "\\}"), URLencode(as.character(`query_id`), reserved = TRUE), url_path)
+
+      # OAuth token
+      header_params["Authorization"] <- paste("Bearer", self$api_client$access_token, sep = " ")
+
+      resp <- self$api_client$CallApi(url = paste0(self$api_client$base_path, url_path),
+                                      method = "GET",
+                                      query_params = query_params,
+                                      header_params = header_params,
+                                      body = body)
+
+      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
+        stream <- arrow::BufferReader$create(httr::content(resp, "raw"))
+        dtype <- rawToChar(stream$Read(1)$data())
+        if (dtype != "A") {
+          # N.B. we assume since the query is a simple select star that we are direct read eligible, if the stack is not
+          # properly configured for direct read this will fail
+          stop("Foundry does not seem to support direct reads, please contact your Palantir administrator", dtype)
```

Comment:
Would be better to phrase this as a configuration issue rather than a property of the Foundry platform